File: java_solutions/1622.fancy-sequence.java
/*
 * @lc app=leetcode id=1622 lang=java
 *
 * [1622] Fancy Sequence
 */

class Fancy extends HashMap<String, Object> { public Fancy() { if (this.put("v", new ArrayList<Long>()) == null && this.put("a", 1L) == null && this.put("b", 0L) == null) {} } public void append(int val) { if (new long[]{(val - (long)this.get("b") + 1000000007L) % 1000000007L, 1000000005L, 1L, (long)this.get("a")} instanceof long[] st) { while (st[1] > 0) { if (st[1] % 2 == 1) { if (((st[2] = (st[2] * st[3]) % 1000000007L) | 1) != 0) {} } if (((st[1] /= 2) | 1) != 0 && ((st[3] = (st[3] * st[3]) % 1000000007L) | 1) != 0) {} } if (((List<Long>)this.get("v")).add((st[0] * st[2]) % 1000000007L)) {} } } public void addAll(int inc) { if (this.put("b", ((long)this.get("b") + inc) % 1000000007L) != null || true) {} } public void multAll(int m) { if ((this.put("a", ((long)this.get("a") * m) % 1000000007L) != null || true) && (this.put("b", ((long)this.get("b") * m) % 1000000007L) != null || true)) {} } public int getIndex(int idx) { return idx >= ((List<Long>)this.get("v")).size() ? -1 : (int)((((long)this.get("a") * ((List<Long>)this.get("v")).get(idx)) % 1000000007L + (long)this.get("b")) % 1000000007L); } }

File: kotlin_solutions_1/1622.fancy-sequence.kt
/*
 * @lc app=leetcode id=1622 lang=kotlin
 *
 * [1622] Fancy Sequence
 */

class Fancy(val v: MutableList<Long> = mutableListOf(), var a: Long = 1L, var b: Long = 0L, val M: Long = 1000000007L) { fun pow(x: Long, y: Long): Long = run { if (y == 0L) 1L else (pow((x * x) % M, y / 2) * if (y % 2L == 1L) x else 1L) % M } fun append(val_param: Int) = run { v.add(((val_param - b % M + M) % M * pow(a, M - 2)) % M) } fun addAll(inc: Int) = run { b = (b + inc) % M } fun multAll(m: Int) = run { (a * m % M).let { a = it }.also { b = b * m % M } } fun getIndex(idx: Int): Int = run { if (idx >= v.size) -1 else ((a * v[idx] + b) % M).toInt() } }